feat(annotations): allow to use annotation of namespace and  pod  to control if disable source ip verification

zbb88888 opened the pull request:
user can set pod annotation to disable source ip verification 

Please ensure your pull request adheres to the following guidelines:

- [x] For first time contributors, read [Submitting a pull request](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#submitting-a-pull-request)
- [x] All code is covered by unit and/or runtime tests where feasible.
- [x] All commits contain a well written commit description including a title,
      description and a `Fixes: #XXX` line if the commit addresses a particular
      GitHub issue.
- [ ] If your commit description contains a `Fixes: <commit-id>` tag, then
      please add the commit author[s] as reviewer[s] to this issue.
- [x] All commits are signed off. See the section [Developer’s Certificate of Origin](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#dev-coo)
- [x] Provide a title or release-note blurb suitable for the release notes.
- [ ] Are you a user of Cilium? Please add yourself to the [Users doc](https://github.com/cilium/cilium/blob/main/USERS.md)
- [ ] Thanks for contributing!

<!-- Description of change -->

Fixes: #issue-number

```release-note
Allow users to disable source IP verification per-pod by setting an annotation. This requires an annotation on the pod's Namespace as well.
```

pchaigno reviewed (changes requested):
I agreed with Casey about the security concern here.

What's the use case for this? It's already possible to disable source IP verification on a per-pod basis via runtime configs AFAIK.

squeed commented on pkg/k8s/watchers/pod.go:
```diff
@@ -393,7 +397,7 @@ func (k *K8sPodWatcher) updateK8sPodV1(ctx context.Context, oldK8sPod, newK8sPod
 
 		if annotationsChanged {
-			if annoChangedBandwidth {
+			if annoChangedBandwidth || annoChangedPriority {
```
Whoops, you found a bug here :-).

Any chance you could make this one line a separate commit for easy backporting?

zbb88888 replied:
@squeed  Hi， I will make a  separate commit PR firstly！

zbb88888 replied:
https://github.com/cilium/cilium/pull/44329 new pr here, I will drop this from this PR

zbb88888 replied:
done

squeed reviewed (changes requested):
Looks good.

One last thing I would like to see: can you move `Documentation/network/pod-mac-address.rst` to `pod-annotations.rst` and document this annotation as well?

squeed reviewed (approved):
Nice work!

pchaigno reviewed (changes requested):
Sorry for the delay to review! 🙇 

The changes look good to me, but I think the commit history needs work. Could you move the documentation update to its own commit(s), maybe with a first commit doing the renaming so git can track it properly? Could you also expend the commit descriptions to explain what you're doing and *why* it's required?

zbb88888 committed:
docs: rename pod-mac-address.rst to pod-annotations.rst

Rename the pod-level annotation documentation file from
pod-mac-address.rst to pod-annotations.rst to serve as a unified
page for all pod annotation features. Update index.rst to reference
the new file name and add a redirect from the old path.

This is a pure rename with no content changes, so git can track
the file history properly.

Signed-off-by: zbb88888 <jmdxjsjgcxy@gmail.com>

pchaigno reviewed (approved):
Perfect, thanks!

joestringer commented on Documentation/network/pod-annotations.rst:
```diff
@@ -0,0 +1,134 @@
+
+   Disabling source IP verification allows IP spoofing from the pod.
+   Only enable this for trusted workloads. Use Kubernetes RBAC to restrict
+   who can modify namespace annotations.
```
Small suggestion to make the risk even more clear:

```suggestion
   Disabling source IP verification allows IP spoofing from the pod.
   This means that the Pod has the capability to bypass network policy
   enforcement within the cluster. Only enable this for trusted workloads.
   Use Kubernetes RBAC to restrict who can modify namespace annotations.
```

zbb88888 replied:
fixed in a new commit

joestringer commented on Documentation/network/pod-annotations.rst:
```diff
@@ -0,0 +1,134 @@
+===========
+
+Namespace annotation (administrator-controlled):
+  ``config.cilium.io/allow-disable-source-ip-verification``
```
The annotation name on one hand makes sense here, but also seems a bit confusing ("allow disable").

I wonder if this could be more clearly expressed with another name, such as:

```suggestion
  ``config.cilium.io/delegate-source-ip-verification``
```

This would mean that the source IP verification behavior is delegated to the namespace owner.

zbb88888 replied:
fixed in a new commit

pchaigno replied:
Let's move these two fixes into the commit they fix.

zbb88888 replied:
@pchaigno Hi, thanks for the review! I have split the commt 4 into commit 2 and commit 3. Please reivew again!

zbb88888 committed:
docs: document disable-source-ip-verification annotation

Add documentation for the new per-pod source IP verification control
annotations to the unified pod-annotations.rst page:

- config.cilium.io/delegate-source-ip-verification (namespace)
- config.cilium.io/disable-source-ip-verification (pod)

The two-layer annotation model requires a namespace-level delegation
from a cluster administrator before any pod can disable source IP
verification. This prevents untrusted workloads from spoofing source
IPs while allowing legitimate use cases like NAT gateways, VPN
endpoints, and load balancers.

Also reorganize the page structure: rename the top-level title to
"Pod Annotations" and nest the existing MAC address section under it.

Signed-off-by: zbb88888 <jmdxjsjgcxy@gmail.com>

zbb88888 committed:
feat(annotations): allow per-pod source IP verification control

Add a two-layer annotation mechanism to control source IP verification
on a per-pod basis, addressing the need for workloads like NAT gateways,
VPN endpoints, and load balancers that must send packets with non-pod
source IPs.

Security model:
  1. Namespace annotation (admin-controlled):
     config.cilium.io/delegate-source-ip-verification
     Delegates source IP verification control to the namespace owner.
     Must be set to a truthy value by a cluster administrator before
     any pod annotation takes effect.
  2. Pod annotation:
     config.cilium.io/disable-source-ip-verification
     Disables source IP verification for the specific pod, but only
     when the namespace permits it.

This design prevents unprivileged users from spoofing IPs while
providing a Kubernetes-native, declarative alternative to the
ephemeral cilium endpoint config CLI.

Implementation:
- pkg/annotation/k8s.go: Define annotation constants.
- pkg/endpoint/endpoint.go: ApplySourceIPVerificationFromAnnotation()
  evaluates namespace + pod annotations and updates the endpoint option.
- pkg/endpoint/metadata/k8s_metadatafetcher.go: Propagate namespace
  annotations into K8sMetadata so metadataResolver can apply SIP
  settings during endpoint creation / restore.
- pkg/endpointmanager/namespace_updater.go: Watch namespace annotation
  changes via statedb Changes() iterator (replacing AllWatch), re-evaluate
  all endpoints when the delegation setting changes, and clean up tracking
  maps on namespace deletion.
- pkg/k8s/watchers/pod.go: React to pod annotation changes in the
  updateK8sPodV1 path, look up namespace annotations, and trigger
  datapath regeneration when the SIP setting changes.

Testing:
- pkg/endpoint/endpoint_test.go: Table-driven tests for all annotation
  combinations, namespace permission gating, and global-default reset.
- pkg/endpointmanager/namespace_updater_test.go: Namespace delete cleanup,
  recreate, and multi-namespace isolation tests.

Signed-off-by: zbb88888 <jmdxjsjgcxy@gmail.com>

joestringer reviewed (approved):
Docs and usage look good to me, thanks!
